databricks/koalas/datetime.py

```diff
@@ -0,0 +1,154 @@
+            return Series(
+                scol, anchor=self._data._kdf, index=self._data._index_info)
+        return wrapper
+    return decorator
+
+
+def defer_to_spark(output_type):
+    """Wraps a function that operates on spark.sql.Column to work with DatetimeMethods.
+
+    The origin function operates on spark.sql.Column and the wrapped function
+    operates on DatetimeMethods.
+    """
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(self, *args):
+            scol = func(self._data._scol, *args).cast(output_type)
+            return Series(scol, anchor=self._data._kdf, index=self._data._index_info)
+        return wrapper
+    return decorator
+
+
+class DatetimeMethods:
+    """Date/Time methods for Koalas Series"""
+    def __init__(self, series):
+        if not isinstance(series.spark_type, (DateType, TimestampType)):
+            raise ValueError(
+                "Cannot call DatetimeMethods on type {}"
+                .format(series.spark_type))
+        self._data = series
+
+    # Properties
+    @lazy_property
+    @defer_to_pandas(output_type=DateType())
+    def date(s):
+        return s.dt.date
+
+    @lazy_property
+    def time(self):
+        raise NotImplementedError()
+
+    @lazy_property
+    def timetz(self):
+        raise NotImplementedError()
+
+    @lazy_property
+    @defer_to_spark(output_type=LongType())
+    def year(col):
+        return F.year(col)
+
+    @lazy_property
+    @defer_to_spark(output_type=LongType())
+    def month(col):
+        return F.month(col)
+
+    @lazy_property
+    @defer_to_spark(output_type=LongType())
+    def day(col):
+        return F.dayofmonth(col)
+
+    @lazy_property
+    @defer_to_spark(output_type=LongType())
+    def hour(col):
+        return F.hour(col)
+
+    @lazy_property
+    @defer_to_spark(output_type=LongType())
+    def minute(col):
+        return F.minute(col)
+
+    @lazy_property
+    @defer_to_spark(output_type=LongType())
+    def second(col):
+        return F.second(col)
+
+    @lazy_property
+    @defer_to_spark(output_type=LongType())
+    def millisecond(col):
+        return F.millisecond(col)
+
+    @lazy_property
+    @defer_to_pandas(output_type=LongType())
+    def microsecond(s):
+        return s.dt.microsecond
+
+    @lazy_property
+    def nanosecond(self):
+        raise NotImplementedError()
+
+    @lazy_property
+    @defer_to_spark(output_type=LongType())
+    def week(col):
+        return F.weekofyear(col)
+
+    @lazy_property
+    @defer_to_spark(output_type=LongType())
+    def weekofyear(col):
+        return F.weekofyear(col)
+
+    @lazy_property
+    @defer_to_pandas(output_type=LongType())
+    def dayofweek(s):
+        return s.dt.dayofweek
+
+    @lazy_property
+    @defer_to_spark(output_type=LongType())
+    def dayofyear(col):
+        return F.dayofyear(col)
+
+    # Methods
+    @defer_to_pandas(output_type=StringType())
+    def strftime(s, date_format):
```

Comment:
The decorator supports multiple arguments, but the first argument must be the "parent" series of the DatetimeMethods object.